Using random ports from portbindings

##### Issue Overview

As user of arquillian cube, I would like to be able to use random ports selected by the docker daemon


##### Current Behaviour
The default port is taken since no remapping is done if cube configuration is like (Cube format):
<property name="dockerContainers">
   containerName_*:
      image: myimage
      portBindings: [9990/tcp,8080/tcp]
</property>

This will try to launch the Docker container with a configuration which is like:
containerName_42953_42954_806c5a61-ec20-4d3a-9073-8e7f5a58b097:
    image: myimage
    killContainer: false
    manual: false
    portBindings: [53607->9990/tcp, 53752->8080/tcp]
    readonlyRootfs: false
    removeVolumes: true

After which the port bindings are not passed to the test-code. Which tries to use 9990/8080. This, of course, doesn't succeed.

##### Expected Behaviour
When running arquillian-cube with random ports, those ports are remapped to the containerConfiguration, so that the tests can run in parallel. So after launching the container(probably event BeforeSuite?) the ports should be rebound if the containerName contains a wildcard.

##### Additional Information

If I want to implement this feature, what would be the way to go. Is this possible to do so and If it is possible, are there some pointers to which information is available where?

Yes. If I manually assign a portbinding it connects. So this works:
containerName_*:
image: myimage
portBindings: [53607->9990/tcp, 53752->8080/tcp]

With the random port bindings, I can see it tries to connect to 9990 using wireshark.

The stacktrace is
Stacktrace:
org.jboss.arquillian.container.spi.client.container.DeploymentException: Cannot deploy: customer-test.war
	at org.jboss.as.arquillian.container.ArchiveDeployer.deployInternal(ArchiveDeployer.java:141)
	at org.jboss.as.arquillian.container.ArchiveDeployer.deployInternal(ArchiveDeployer.java:121)
	at org.jboss.as.arquillian.container.ArchiveDeployer.deploy(ArchiveDeployer.java:83)
	at org.jboss.as.arquillian.container.CommonDeployableContainer.deploy(CommonDeployableContainer.java:236)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$3.call(ContainerDeployController.java:161)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$3.call(ContainerDeployController.java:128)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.executeOperation(ContainerDeployController.java:271)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.deploy(ContainerDeployController.java:127)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:96)
	at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:99)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:81)
	at org.jboss.arquillian.container.impl.client.container.DeploymentExceptionHandler.verifyExpectedExceptionDuringDeploy(DeploymentExceptionHandler.java:50)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:96)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:88)
	at org.jboss.arquillian.container.impl.client.ContainerDeploymentContextHandler.createDeploymentContext(ContainerDeploymentContextHandler.java:78)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:96)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:88)
	at org.jboss.arquillian.container.impl.client.ContainerDeploymentContextHandler.createContainerContext(ContainerDeploymentContextHandler.java:57)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:96)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:88)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:145)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:116)
	at org.jboss.arquillian.core.impl.EventImpl.fire(EventImpl.java:67)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$1.perform(ContainerDeployController.java:95)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$1.perform(ContainerDeployController.java:80)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.forEachDeployment(ContainerDeployController.java:263)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.forEachManagedDeployment(ContainerDeployController.java:239)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.deployManaged(ContainerDeployController.java:79)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:96)
	at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:99)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:81)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:145)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:116)
	at org.jboss.arquillian.core.impl.EventImpl.fire(EventImpl.java:67)
	at org.jboss.arquillian.container.test.impl.client.ContainerEventController.execute(ContainerEventController.java:101)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:96)
	at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:99)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:81)
	at org.jboss.arquillian.test.impl.TestContextHandler.createClassContext(TestContextHandler.java:92)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:96)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:88)
	at org.jboss.arquillian.test.impl.TestContextHandler.createSuiteContext(TestContextHandler.java:73)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:96)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:88)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:145)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:116)
	at org.jboss.arquillian.test.impl.EventTestRunnerAdaptor.beforeClass(EventTestRunnerAdaptor.java:87)
	at org.jboss.arquillian.junit.Arquillian$2.evaluate(Arquillian.java:202)
	at org.jboss.arquillian.junit.Arquillian.multiExecute(Arquillian.java:431)
	at org.jboss.arquillian.junit.Arquillian.access$200(Arquillian.java:55)
	at org.jboss.arquillian.junit.Arquillian$3.evaluate(Arquillian.java:219)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.jboss.arquillian.junit.Arquillian.run(Arquillian.java:167)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:83)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:74)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at org.junit.platform.console.tasks.ConsoleTestExecutor.executeTests(ConsoleTestExecutor.java:65)
	at org.junit.platform.console.tasks.ConsoleTestExecutor.lambda$execute$0(ConsoleTestExecutor.java:57)
	at org.junit.platform.console.tasks.CustomContextClassLoaderExecutor.invoke(CustomContextClassLoaderExecutor.java:33)
	at org.junit.platform.console.tasks.ConsoleTestExecutor.execute(ConsoleTestExecutor.java:57)
	at org.junit.platform.console.ConsoleLauncher.executeTests(ConsoleLauncher.java:84)
	at org.junit.platform.console.ConsoleLauncher.execute(ConsoleLauncher.java:74)
	at org.junit.platform.console.ConsoleLauncher.execute(ConsoleLauncher.java:47)
	at org.junit.platform.console.ConsoleLauncher.main(ConsoleLauncher.java:39)
Caused by: java.net.ConnectException: Connection refused
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)
	at org.xnio.nio.WorkerThread$ConnectHandle.handleReady(WorkerThread.java:319)
	at org.xnio.nio.WorkerThread.run(WorkerThread.java:559)
	at ...asynchronous invocation...(Unknown Source)
	at org.jboss.remoting3.EndpointImpl.doConnect(EndpointImpl.java:294)
	at org.jboss.remoting3.EndpointImpl.doConnect(EndpointImpl.java:275)
	at org.jboss.remoting3.EndpointImpl.connect(EndpointImpl.java:379)
	at org.jboss.remoting3.EndpointImpl.connect(EndpointImpl.java:367)
	at org.jboss.as.protocol.ProtocolConnectionUtils.connect(ProtocolConnectionUtils.java:83)
	at org.jboss.as.protocol.ProtocolConnectionUtils.connectSync(ProtocolConnectionUtils.java:114)
	at org.jboss.as.protocol.ProtocolConnectionManager$EstablishingConnection.connect(ProtocolConnectionManager.java:257)
	at org.jboss.as.protocol.ProtocolConnectionManager.connect(ProtocolConnectionManager.java:71)
	at org.jboss.as.protocol.mgmt.FutureManagementChannel$Establishing.getChannel(FutureManagementChannel.java:212)
	at org.jboss.as.controller.client.impl.RemotingModelControllerClient.getOrCreateChannel(RemotingModelControllerClient.java:146)
	at org.jboss.as.controller.client.impl.RemotingModelControllerClient$1.getChannel(RemotingModelControllerClient.java:65)
	at org.jboss.as.protocol.mgmt.ManagementChannelHandler.executeRequest(ManagementChannelHandler.java:147)
	at org.jboss.as.protocol.mgmt.ManagementChannelHandler.executeRequest(ManagementChannelHandler.java:122)
	at org.jboss.as.controller.client.impl.AbstractModelControllerClient.executeRequest(AbstractModelControllerClient.java:263)
	at org.jboss.as.controller.client.impl.AbstractModelControllerClient.execute(AbstractModelControllerClient.java:168)
	at org.jboss.as.controller.client.impl.AbstractModelControllerClient.executeAsync(AbstractModelControllerClient.java:111)
	at org.jboss.as.controller.client.helpers.standalone.impl.ModelControllerClientServerDeploymentManager.executeOperation(ModelControllerClientServerDeploymentManager.java:50)
	at org.jboss.as.controller.client.helpers.standalone.impl.AbstractServerDeploymentManager.execute(AbstractServerDeploymentManager.java:79)
	at org.jboss.as.controller.client.helpers.standalone.ServerDeploymentHelper.deploy(ServerDeploymentHelper.java:54)
	at org.jboss.as.arquillian.container.ArchiveDeployer.deployInternal(ArchiveDeployer.java:135)
	... 110 more

--- a/core/src/main/java/org/arquillian/cube/impl/client/CubeExtension.java
+++ b/core/src/main/java/org/arquillian/cube/impl/client/CubeExtension.java
@@ -1,6 +1,7 @@
 package org.arquillian.cube.impl.client;
 
 import org.arquillian.cube.impl.client.container.ContainerConfigurationController;
+import org.arquillian.cube.impl.client.container.ContainerMetadataUpdater;
 import org.arquillian.cube.impl.client.container.CubeContainerLifecycleController;
 import org.arquillian.cube.impl.client.container.CubeRemoteCommandObserver;
 import org.arquillian.cube.impl.client.container.ProtocolMetadataUpdater;
@@ -38,7 +39,8 @@ public class CubeExtension implements LoadableExtension {
         // Arquillian Container integration
         // Only register if container-test-spi is on classpath
         if (Validate.classExists("org.jboss.arquillian.container.test.spi.client.deployment.AuxiliaryArchiveAppender")) {
-            builder.observer(ProtocolMetadataUpdater.class)
+            builder.observer(ContainerMetadataUpdater.class)
+                .observer(ProtocolMetadataUpdater.class)
                 .observer(CubeContainerLifecycleController.class)
                 .observer(ContainerConfigurationController.class)
                 .observer(CubeRemoteCommandObserver.class);

--- a/core/src/main/java/org/arquillian/cube/impl/client/container/ContainerConfigurationController.java
+++ b/core/src/main/java/org/arquillian/cube/impl/client/container/ContainerConfigurationController.java
@@ -106,13 +106,7 @@ public class ContainerConfigurationController {
                 return -1;
             }
             return (int) method.invoke(configurationInstance);
-        } catch (SecurityException e) {
-            throw new IllegalArgumentException(e);
-        } catch (IllegalAccessException e) {
-            throw new IllegalArgumentException(e);
-        } catch (IllegalArgumentException e) {
-            throw new IllegalArgumentException(e);
-        } catch (InvocationTargetException e) {
+        } catch (SecurityException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
             throw new IllegalArgumentException(e);
         }
     }

--- a/core/src/main/java/org/arquillian/cube/impl/client/container/ContainerMetadataUpdater.java
+++ b/core/src/main/java/org/arquillian/cube/impl/client/container/ContainerMetadataUpdater.java
@@ -1,0 +1,33 @@
+package org.arquillian.cube.impl.client.container;
+
+import java.lang.reflect.Field;
+import java.util.Map;
+import java.util.logging.Logger;
+import org.arquillian.cube.impl.model.ContainerMetadata;
+import org.jboss.arquillian.container.spi.Container;
+import org.jboss.arquillian.container.spi.ContainerRegistry;
+import org.jboss.arquillian.core.api.annotation.Observes;
+
+public class ContainerMetadataUpdater {
+
+    private static final Logger log = Logger.getLogger(ContainerMetadataUpdater.class.getName());
+
+    public void updateResolvedContainersName(@Observes ContainerMetadata containerMetadata,
+        ContainerRegistry containerRegistry) {
+        final Map<String, String> containers = containerMetadata.getNames();
+        for (Container container : containerRegistry.getContainers()) {
+            if (containers.containsKey(container.getName())) {
+                final String mappedContainerName = containers.get(container.getName());
+                try {
+                    Field name = container.getClass().getDeclaredField("name");
+                    name.setAccessible(true);
+                    name.set(container, mappedContainerName);
+                } catch (NoSuchFieldException | IllegalAccessException e) {
+                    log.warning(
+                        String.format("Failed to update name %s for container %s", mappedContainerName,
+                            container.getName()));
+                }
+            }
+        }
+    }
+}

--- a/core/src/main/java/org/arquillian/cube/impl/model/ContainerMetadata.java
+++ b/core/src/main/java/org/arquillian/cube/impl/model/ContainerMetadata.java
@@ -1,0 +1,16 @@
+package org.arquillian.cube.impl.model;
+
+import java.util.Map;
+
+public class ContainerMetadata {
+
+    private final Map<String, String> names;
+
+    public ContainerMetadata(Map<String, String> names) {
+        this.names = names;
+    }
+
+    public Map<String, String> getNames() {
+        return names;
+    }
+}

--- a/docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/CubeDockerConfigurator.java
+++ b/docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/CubeDockerConfigurator.java
@@ -1,5 +1,9 @@
 package org.arquillian.cube.docker.impl.client;
 
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
 import org.arquillian.cube.HostIpContext;
 import org.arquillian.cube.docker.impl.client.config.CubeContainer;
 import org.arquillian.cube.docker.impl.client.config.DockerCompositions;
@@ -8,10 +12,10 @@ import org.arquillian.cube.docker.impl.client.config.StarOperator;
 import org.arquillian.cube.docker.impl.util.Boot2Docker;
 import org.arquillian.cube.docker.impl.util.DefaultDocker;
 import org.arquillian.cube.docker.impl.util.DockerMachine;
-import org.arquillian.cube.docker.impl.util.OperatingSystem;
 import org.arquillian.cube.docker.impl.util.OperatingSystemInterface;
 import org.arquillian.cube.docker.impl.util.OperatingSystemResolver;
 import org.arquillian.cube.docker.impl.util.Top;
+import org.arquillian.cube.impl.model.ContainerMetadata;
 import org.arquillian.cube.spi.CubeConfiguration;
 import org.jboss.arquillian.config.descriptor.api.ArquillianDescriptor;
 import org.jboss.arquillian.core.api.Injector;
@@ -21,19 +25,11 @@ import org.jboss.arquillian.core.api.annotation.ApplicationScoped;
 import org.jboss.arquillian.core.api.annotation.Inject;
 import org.jboss.arquillian.core.api.annotation.Observes;
 
-import java.util.HashMap;
-import java.util.Map;
-import java.util.Optional;
-import java.util.Random;
-import java.util.UUID;
-import java.util.logging.Logger;
-
 public class CubeDockerConfigurator {
 
     public static final String DOCKER_HOST = "DOCKER_HOST";
     private static final String EXTENSION_NAME = "docker";
-    private static Random random = new Random();
-    private static Logger log = Logger.getLogger(CubeDockerConfigurator.class.getName());
+
     @Inject
     @ApplicationScoped
     private InstanceProducer<CubeDockerConfiguration> configurationProducer;
@@ -58,6 +54,10 @@ public class CubeDockerConfigurator {
     @ApplicationScoped
     private InstanceProducer<OperatingSystemInterface> operatingSystemInstanceProducer;
 
+    @Inject
+    @ApplicationScoped
+    private InstanceProducer<ContainerMetadata> containerMetadataInstanceProducer;
+
 
     public void configure(@Observes CubeConfiguration event, ArquillianDescriptor arquillianDescriptor) {
         configure(arquillianDescriptor);
@@ -75,9 +75,7 @@ public class CubeDockerConfigurator {
 
         Map<String, String> config = arquillianDescriptor.extension(EXTENSION_NAME).getExtensionProperties();
         CubeDockerConfigurationResolver resolver = new CubeDockerConfigurationResolver(topInstance.get(),
-            dockerMachineInstance.get(),
-            boot2DockerInstance.get(),
-            new DefaultDocker(),
+            dockerMachineInstance.get(), boot2DockerInstance.get(), new DefaultDocker(),
             operatingSystemInstanceProducer.get());
         resolver.resolve(config);
         CubeDockerConfiguration cubeConfiguration = CubeDockerConfiguration.fromMap(config, injectorInstance.get());
@@ -109,6 +107,7 @@ public class CubeDockerConfigurator {
             dockerContainersContent.setNetworks(resolvedNetworks);
         }
         final Map<String, CubeContainer> resolvedContainers = new HashMap<>();
+        final Map<String, String> containerMetadata = new HashMap<>();
         final Map<String, CubeContainer> containers = dockerContainersContent.getContainers();
         for (Map.Entry<String, CubeContainer> container : containers.entrySet()) {
 
@@ -126,12 +125,15 @@ public class CubeDockerConfigurator {
 
                 String newId = StarOperator.generateNewName(templateName, uuid);
                 resolvedContainers.put(newId, cubeContainer);
+                containerMetadata.put(templateName, newId);
             } else {
                 resolvedContainers.put(containerId, cubeContainer);
             }
         }
+        if (containerMetadataInstanceProducer != null && !containerMetadata.isEmpty()) {
+            containerMetadataInstanceProducer.set(new ContainerMetadata(containerMetadata));
+        }
         dockerContainersContent.setContainers(resolvedContainers);
-
         return cubeConfiguration;
     }
 

--- a/docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/config/StarOperator.java
+++ b/docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/config/StarOperator.java
@@ -86,11 +86,7 @@ public class StarOperator {
         if (cubeContainer.getNetworks() != null) {
             ArrayList<String> networks = new ArrayList<>();
             for (String network : cubeContainer.getNetworks()) {
-                if (networkResolutions.containsKey(network)) {
-                    networks.add(networkResolutions.get(network));
-                } else {
-                    networks.add(network);
-                }
+                networks.add(networkResolutions.getOrDefault(network, network));
             }
             cubeContainer.setNetworks(networks);
         }

--- a/docker/ftest-container-star-operator/pom.xml
+++ b/docker/ftest-container-star-operator/pom.xml
@@ -1,0 +1,87 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <parent>
+    <artifactId>arquillian-cube-docker-parent</artifactId>
+    <groupId>org.arquillian.cube</groupId>
+    <version>1.15.4-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>arquillian-cube-docker-ftest-container-star-operator</artifactId>
+
+  <properties>
+    <version.javaee_7>1.0.1.Final</version.javaee_7>
+    <version.wildfly>8.2.0.Final</version.wildfly>
+  </properties>
+
+  <dependencyManagement>
+    <dependencies>
+      <dependency>
+        <groupId>org.jboss.spec</groupId>
+        <artifactId>jboss-javaee-7.0</artifactId>
+        <version>${version.javaee_7}</version>
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.jboss.arquillian.junit</groupId>
+      <artifactId>arquillian-junit-container</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.arquillian.cube</groupId>
+      <artifactId>arquillian-cube-docker</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.arquillian.cube</groupId>
+      <artifactId>arquillian-cube-requirement</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.jboss.spec.javax.servlet</groupId>
+      <artifactId>jboss-servlet-api_3.1_spec</artifactId>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.jboss.spec.javax.ejb</groupId>
+      <artifactId>jboss-ejb-api_3.2_spec</artifactId>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.jboss.spec.javax.annotation</groupId>
+      <artifactId>jboss-annotations-api_1.2_spec</artifactId>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.wildfly</groupId>
+      <artifactId>wildfly-arquillian-container-remote</artifactId>
+      <version>${version.wildfly}</version>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-deploy-plugin</artifactId>
+        <configuration>
+          <skip>true</skip>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+</project>

--- a/docker/ftest-container-star-operator/src/main/java/org/arquillian/cube/servlet/HelloWorldServlet.java
+++ b/docker/ftest-container-star-operator/src/main/java/org/arquillian/cube/servlet/HelloWorldServlet.java
@@ -1,0 +1,22 @@
+package org.arquillian.cube.servlet;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/HelloWorld")
+public class HelloWorldServlet extends HttpServlet {
+
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+
+        PrintWriter writer = resp.getWriter();
+        writer.println("Hello World");
+    }
+}

--- a/docker/ftest-container-star-operator/src/test/java/org/arquillian/cube/servlet/HelloWorldServletIT.java
+++ b/docker/ftest-container-star-operator/src/test/java/org/arquillian/cube/servlet/HelloWorldServletIT.java
@@ -1,0 +1,50 @@
+package org.arquillian.cube.servlet;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import org.arquillian.cube.docker.impl.requirement.RequiresDocker;
+import org.arquillian.cube.docker.impl.requirement.RequiresDockerMachine;
+import org.arquillian.cube.requirement.ArquillianConditionalRunner;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.test.api.ArquillianResource;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+@Category({RequiresDocker.class, RequiresDockerMachine.class})
+@RunWith(ArquillianConditionalRunner.class)
+public class HelloWorldServletIT {
+
+    @Deployment(testable = false)
+    public static WebArchive create() {
+        return ShrinkWrap.create(WebArchive.class, "hello.war").addClass(HelloWorldServlet.class);
+    }
+
+    @Test
+    public void should_parse_and_load_configuration_file(@ArquillianResource URL base) throws IOException {
+
+        URL obj = new URL(base, "HelloWorld");
+        HttpURLConnection con = (HttpURLConnection) obj.openConnection();
+        con.setRequestMethod("GET");
+
+        BufferedReader in = new BufferedReader(
+            new InputStreamReader(con.getInputStream()));
+        String inputLine;
+        StringBuffer response = new StringBuffer();
+
+        while ((inputLine = in.readLine()) != null) {
+            response.append(inputLine);
+        }
+        in.close();
+
+        assertThat(response.toString(), is("Hello World"));
+    }
+}

--- a/docker/ftest-container-star-operator/src/test/resources/arquillian.xml
+++ b/docker/ftest-container-star-operator/src/test/resources/arquillian.xml
@@ -1,0 +1,27 @@
+<?xml version="1.0"?>
+<arquillian xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xmlns="http://jboss.org/schema/arquillian"
+  xsi:schemaLocation="http://jboss.org/schema/arquillian
+    http://jboss.org/schema/arquillian/arquillian_1_0.xsd">
+
+  <extension qualifier="docker">
+    <property name="definitionFormat">CUBE</property>
+    <property name="dockerContainers">
+      wildfly*:
+        buildImage:
+          dockerfileLocation: src/test/resources/wildfly
+          noCache: true
+        exposedPorts: [8080/tcp, 9990/tcp]
+        await:
+          strategy: polling
+        portBindings: [63785->9990/tcp, 63786->8080/tcp]
+    </property>
+  </extension>
+
+  <container qualifier="wildfly" default="true">
+    <configuration>
+      <property name="username">admin</property>
+      <property name="password">Admin#70365</property>
+    </configuration>
+  </container>
+</arquillian>

--- a/docker/ftest-container-star-operator/src/test/resources/wildfly/Dockerfile
+++ b/docker/ftest-container-star-operator/src/test/resources/wildfly/Dockerfile
@@ -1,0 +1,3 @@
+FROM jboss/wildfly:10.1.0.Final
+RUN /opt/jboss/wildfly/bin/add-user.sh -up mgmt-users.properties admin Admin#70365 --silent
+CMD ["/opt/jboss/wildfly/bin/standalone.sh", "-b", "0.0.0.0", "-bmanagement", "0.0.0.0"]

--- a/docker/pom.xml
+++ b/docker/pom.xml
@@ -63,6 +63,7 @@
     <module>ftest-system-properties</module>
     <module>ftest-junit-rules</module>
     <module>ftest-docker-junit5</module>
+    <module>ftest-container-star-operator</module>
   </modules>
 
 </project>

--- a/openshift/ftest-standalone/src/test/java/org/arquillian/cube/openshift/standalone/FooTest.java
+++ b/openshift/ftest-standalone/src/test/java/org/arquillian/cube/openshift/standalone/FooTest.java
@@ -1,0 +1,4 @@
+package org.arquillian.cube.openshift.standalone;
+
+public class FooTest {
+}

